pkg/querier/queryrange/volume_cache.go

```diff
@@ -0,0 +1,128 @@
+package queryrange
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type VolumeSplitter struct {
+	cacheKeyLimits
+}
+
+// GenerateCacheKey generates a cache key based on the userID, Request and interval.
+func (i VolumeSplitter) GenerateCacheKey(ctx context.Context, userID string, r queryrangebase.Request) string {
+	cacheKey := i.cacheKeyLimits.GenerateCacheKey(ctx, userID, r)
```

Comment:
This cache key doesn't take into consideration `targetLabels` or `aggregateBy`. I think that means it will return incorrect results if, for example, a user requests the same matchers but a different target label.

Reply:
very good point, thank you!